### ISSUE TITLE
Sandbox is accidentally mounting everything as read-write

### ISSUE THREAD

**Issue by philwo** [bazelbuild/bazel#1364]
So, this is a bit embarrassing and mysterious at the same time:
- Embarrassing: Our sandbox accidentally mounts everything as read-write, because Linux handily ignores MS_RDONLY when you also specify MS_BIND and we don't have tests for this.
- Mysterious: Why did nobody ever notice this?

```
cd ~/workspace

philwo@philwo:~/src/bazel (master)$ sudo mkdir /etc/worldwrite; sudo chmod 0777 /etc/worldwrite
philwo@philwo:~/src/bazel (master)$ echo "This is readonly." > readonly.txt

philwo@philwo:~/src/bazel (master)$ cat >BUILD <<'EOF'
genrule(
	name = "owned",
	srcs = ["readonly.txt"],
	outs = ["test.txt"],
	cmd = (
		"touch /etc/worldwrite/OWNED;" +
		"echo \"This is not readonly, this is SPARTA!\" > $(location :readonly.txt);" +
		"touch \"$@\""
	)
)
EOF

philwo@philwo:~/src/bazel (master)$ bazel build --spawn_strategy=sandboxed --genrule_strategy=sandboxed :owned
INFO: Found 1 target...
Target //examples:sandbox up-to-date:
  bazel-genfiles/examples/test.txt
INFO: Elapsed time: 0.269s, Critical Path: 0.06s

philwo@philwo:~/src/bazel (master)$ ls -l /etc/worldwrite/OWNED 
-rw-r--r-- 1 philwo eng 0 Jun 10 14:28 /etc/worldwrite/OWNED

philwo@philwo:~/src/bazel (master)$ cat examples/readonly.txt 
This is not readonly, this is SPARTA!
```

This means, that a sandboxed action can a) influence the host environment and b) modify the source-code that you're building.

I'm working on a fix now.

**Comment by philwo** [comment]
FYI @bsilver8192 @AustinSchuh - please take care when relying on sandboxing until this is fixed.

**Comment by bsilver8192** [comment]
Thanks for the heads-up.

I think part of the reason I never noticed this before is because when I accidentally try overwriting rule inputs, they're often outputs of other rules which are read-only anyways so it still fails. Still a pretty exciting bug though...

**Comment by philwo** [comment]
I had to rollback the fix, because Ubuntu 15.10 seems to have an additional security check in the kernel that disallows loading shared libraries from mountpoints that have noexec set.

I'll need to come up with a better way to detect when to use noexec and when not.

**Comment by philwo** [comment]
Fix (as part of "Sandbox 2.0") sent for code-review: https://bazel-review.googlesource.com/#/c/4053/
